File: transposeMatrix.js
const print = console.table;

function transpose(mat) {
  let n_rows = mat.length;
  let n_cols = mat[0].length;
  let t_mat = [];
  for(let i = 0; i < n_cols; i++) {
    t_mat.push([]);
  }

  for(let i = 0; i < n_rows; i++) {
    for(let j = 0; j < n_cols; j++) {
      t_mat[j][i] = mat[i][j];
    }
  }
  return t_mat;
}

const mat = [
  [1, 2, 3],
  [4, 5, 6]
];

const mat1 = [
  [1, 4],
  [2, 5],
  [3, 6]
];

print(transpose(mat));
print(transpose(mat1));

File: transposeMatrix.cpp
#include <bits/stdc++.h>
using namespace std;

vector<vector<int>> transpose(vector<vector<int>>& matrix) {
  long long n_rows = matrix.size();
  long long n_cols = matrix[0].size();
  vector<vector<int>> res(n_cols, vector<int>(n_rows));

  for(size_t i = 0; i < n_rows; i++) {
    for(size_t j = 0; j < n_cols; j++) {
      res[j][i] = matrix[i][j];
    }
  }
  return res;
}

void printMatrix(vector<vector<int>>&& matrix) {
  for(auto row : matrix) {
    for(int num : row) {
      cout << num << " ";
    }
    cout << endl;
  }
}

int main() {
  vector<vector<int>> matrix = {
    {1, 2, 3},
    {4, 5, 6},
    {7, 8, 9}
  };
  printMatrix(move(matrix));
  cout << "*****" << endl;
  printMatrix(transpose(matrix));
  return 0;
}